tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -787,9 +788,87 @@ class SPIRVEmitter : public ASTConsumer {
     }
   }
 
+  uint32_t doIntrinsicCallExpr(const CallExpr *callExpr) {
+    const FunctionDecl *callee = callExpr->getDirectCallee();
+    const uint32_t returnType =
+        typeTranslator.translateType(callExpr->getType());
+
+    assert(hlsl::IsIntrinsicOp(callee) &&
+           "doIntrinsicCallExpr was called for a non-intrinsic function.");
+
+    // Figure out which intrinsic function to translate.
+    llvm::StringRef group;
+    uint32_t opcode;
+    hlsl::GetIntrinsicOp(callee, opcode, group);
+
+    switch (static_cast<hlsl::IntrinsicOp>(opcode)) {
+    case hlsl::IntrinsicOp::IOP_dot: {
+      // Get the function parameters.
+      assert(callExpr->getNumArgs() == 2u &&
+             "dot function may only take 2 parameters.");
+      const Expr *arg0 = callExpr->getArg(0);
+      const Expr *arg1 = callExpr->getArg(1);
+      const uint32_t id0 = doExpr(arg0);
+      const uint32_t id1 = doExpr(arg1);
+      QualType t0 = arg0->getType();
+      QualType t1 = arg1->getType();
+      const size_t vec0Size = hlsl::GetHLSLVecSize(t0);
+      const size_t vec1Size = hlsl::GetHLSLVecSize(t1);
+      const QualType vec0ComponentType = hlsl::GetHLSLVecElementType(t0);
+      const QualType vec1ComponentType = hlsl::GetHLSLVecElementType(t1);
+
+      assert(callExpr->getType() == vec1ComponentType &&
```

Comment:
I'll leave the asserts in. It seems that if the vector sizes don't match, DXC issues a warning - not an error. It truncates the larger vector to the size of the smaller vector (which is sad & bad).

In case of type mismatch, there isn't even a warning!